com.unity.render-pipelines.high-definition/Runtime/Debug/DebugViewTiles.shader

```diff
@@ -50,7 +50,7 @@ Shader "Hidden/HDRP/DebugViewTiles"
             int _ClusterDebugMode;
 
             StructuredBuffer<uint> g_TileList;
-            Buffer<uint> g_DispatchIndirectBuffer;
+            StructuredBuffer<uint> g_DispatchIndirectBuffer; // Has to be a StructuredBuffer (not Buffer) to work with DXC https://github.com/KhronosGroup/SPIRV-Cross/issues/1362
```

Comment:
My memory is fuzzy, but I guess there was a reasons to use Buffer instead of StructuredBuffer all over the places. Guess StructuredBuffer was not working on some platforms or in some context. Anyone remember? @mmikk @jessebarker @FrancescoC-unity @EvgeniiG

Reply:
(This comment is for the whole PR, not this particular line)

Reply:
@sebastienlagarde I see that a bunch of these `Indirect arguments have to be in a _buffer_, not a structured buffer` were added by Rune 4 years ago. But why they have to be in buffer? Maybe they had to be in buffer back then, but since then some code got tweaked/fixed and they no longer have to?

Reply:
I don't recall any specific issues with Buffer vs StructuredBuffer and their translations to MSL or GLSL/SPIR-V.  With Metal, we've typically had more issues based on data type or pixel format rather than how they're organized (apart from issues around texture atomics).  The failures I'm seeing on Yamato don't seem related to shaders and there are a lot of passing suites for Vulkan and macOS, so I'm guessing this is ok.

Reply:
Structured buffers that are not 128-bit aligned have terrible performance on Nvidia. Not sure about AMD.

Reply:
This source change is only needed to generate correct MSL.

Not sure if the change is harmful for Metal with the old compiler though. When compiling this to MSL with the old compiler backend (FXC+HLSLcc) the output changes from:
`const device uint *TestRWBuffer [[ buffer(0) ]]`
to
```
struct TestRWBuffer_Type {    uint value[1]; };
const device TestRWBuffer_Type *TestRWBuffer [[ buffer(0) ]] // << in the entry point argument list
```
so my guess would be that this change doesn't have perf implication on Metal as this is still a pointer to an array of uints in the end ( correct me if I'm wrong as I'm just guessing ).

Do you think we could maybe use macros for [RW]Buffer which would only be defined to [RW]StructuredBuffer on DXC+Metal (or maybe Metal with all compilers?) to avoid use of unaligned StructuredBuffer on other APIs?

@EvgeniiG What is our current minimal buffer alignment, do we often use buffers not aligned to 128B?

Reply:
In the current shader file, the size of the structure is sizeof(unit), so it is not 128-bit aligned.

A bigger concern (which is possibly what Rune referred to) is that I don't believe there is a guarantee about the exact memory layout of structured buffers, so interpreting it as a regular array of float3s or float4s (which is what an indirect dispatch does) may potentially be unsafe.

_"Structured identifies the Resource as a single contiguous block of memory with no mipmaps, nor array slices, but it does have a structure size (stride), so that it represents an array of structures. Implementations can take advantage of knowing there is a fixed structure size in the way they lay out the memory physically (hidden from the application)."_

https://microsoft.github.io/DirectX-Specs/d3d/archive/D3D11_3_FunctionalSpec.htm#StructuredBuffer

Reply:
Sorry for the delay, had been attending to other PRs the past week.

From what I hear the main concern is with the change to StructuredBuffers are:
1) Using them for Indirect buffers for D3D APIs could _potentially_ not work on some hardware where SBs don't have an array of structures memory layout as that's something the d3d11 spec does not enforce
2) Structured buffers can be less performant on NVIDIA hardware than regular buffers

I could either:
1) Leave the code as is
2) Modify the changes I did to indirect buffers in particular by adding some macro like INDIRECT_BUFFER and RW_INDIRECT_BUFFER, which would only be  defined to StructuredBuffer and RWStructured for the combo of DXC+Metal (as the above 2 concerns are for D3D or NVIDIA, neither of which concern Metal)
4) Some other suggestion

So @sebastienlagarde, @EvgeniiG what's your verdict on what should be done?

Reply:
Ok, so I randomly had an epiphany, that this replacement can be done with the shader compiler preprocessor with `-DRWBuffer=RWStructuredBuffer` `-DBuffer=StructuredBuffer` without needing to add ad-hoc macros or change shader source.

I undid the manual shader changes in f9697a37a717e65cb80cb27c0aaa78c3cea583d4.